### ISSUE TITLE
With 128-bit capabilities capretaddr is failing

### ISSUE THREAD

**Issue by staceyson** [CTSRD-CHERI/cheri-c-tests#4]
Let's separate this out as its own issue...

`capretaddr` is failing with 128-bit magic capabilities:

```
 # ./capretaddr 
ret = { p:78017 b:0000000000000000 l:0000010000000000 o:1200977dc }
testfn = { p:78017 b:0000000000000000 l:0000010000000000 o:1200977f8 }
Assertion failed: (__builtin_memcap_offset_get(ret) > __builtin_memcap_offset_get(testfn)), function foo, file capretaddr.c, line 53.
```

The capability for `testfn` seems correct:

```
00000001200977f8 <testfn>:
   1200977f8:   67bdffc0        daddiu  sp,sp,-64
   1200977fc:   ebcbe9c3        csd     s8,sp,56(c11)
   120097800:   fa2be820        csc     c17,sp,32(c11)
```

The capability for the `ret` points to the return address for the function call in `main`:

```
0000000120097780 <main>:
   120097780:   67bdffc0        daddiu  sp,sp,-64
   120097784:   ebcbe9c3        csd     s8,sp,56(c11)
   120097788:   eb8be983        csd     gp,sp,48(c11)
   12009778c:   fa2be820        csc     c17,sp,32(c11)
   120097790:   03a0f025        move    s8,sp
   120097794:   6404ffe0        daddiu  a0,zero,-32
   120097798:   49b96002        cgetoffset      t9,c12
   12009779c:   3c010003        lui     at,0x3
   1200977a0:   48006007        cgetpcc c12
   1200977a4:   0039082d        daddu   at,at,t9
   1200977a8:   643c0c40        daddiu  gp,at,3136
   1200977ac:   3c010000        lui     at,0x0
   1200977b0:   003c082d        daddu   at,at,gp
   1200977b4:   dc219430        ld      at,-27600(at)
   1200977b8:   49ac6041        csetoffset      c12,c12,at
   1200977bc:   48f16000        cjalr   c17,c12
   1200977c0:   03a4e824        and     sp,sp,a0
   1200977c4:   df819038        ld      at,-28616(gp)
   1200977c8:   48006007        cgetpcc c12
   1200977cc:   64217848        daddiu  at,at,30792
   1200977d0:   49ac6041        csetoffset      c12,c12,at
   1200977d4:   48f16000        cjalr   c17,c12
   1200977d8:   00000000        nop
   1200977dc:   24020000        li      v0,0              <---  ret   (???)
   1200977e0:   03c0e825        move    sp,s8
   1200977e4:   da2be820        clc     c17,sp,32(c11)
   1200977e8:   cb8be983        cld     gp,sp,48(c11)
   1200977ec:   cbcbe9c3        cld     s8,sp,56(c11)
   1200977f0:   49008800        cjr     c17
   1200977f4:   67bd0040        daddiu  sp,sp,64

00000001200977f8 <testfn>:
```

The address of `foo` is:

```
0000000120097848 <foo>:
 */
#include "cheri_c_test.h"

static void testfn(void);
static void foo(void)
{
   120097848:   67bdff60        daddiu  sp,sp,-160
   12009784c:   67a40098        daddiu  a0,sp,152
```

Looking at the instruction trace:

```
0x00000001200977c4:  ld at,-28616(gp)
    Memory Read [00000001200c13f8] = 0000000120090000
    Write at = 0000000120090000
0x00000001200977c8:  cgetpcc    c12
    Write C12|v:1 s:0 p:00078017 b:0000000000000000 l:0000010000000000
             |o:00000001200977c8 t:0
0x00000001200977cc:  daddiu     at,at,30792
    Write at = 0000000120097848
0x00000001200977d0:  csetoffset c12,c12,at
    Write C12|v:1 s:0 p:00078017 b:0000000000000000 l:0000010000000000
             |o:0000000120097848 t:0        <====== foo() is called from main() and not testfn()
0x00000001200977d4:  cjalr      c17,c12
    Write C17|v:1 s:0 p:00078017 b:0000000000000000 l:0000010000000000
             |o:00000001200977dc t:0
0x00000001200977d8:  nop
```

It seems the optimizer is messing up this test?

**Comment by staceyson** [comment]
The `capretaddr` test passes if compiled with `-O0` instead of `-O3` optimization.

**Comment by davidchisnall** [comment]
I believe that this was passing because a side effect of `-mabi=sandbox` was to break the inliner.  This is now fixed, which allows the function to be inlined.  Sticking `__attribute__((noinline))` on `testfn` ought to prevent it from being inlined into `main` and allow the test to pass again.

**Comment by staceyson** [comment]
Yes, adding `__attribute__((noinline))` to `testify` fixes this.  Unfortunately, I don't have permission to update this repository and can't submit the change.

**Comment by davidchisnall** [comment]
Your access should be fixed now.

**Comment by staceyson** [comment]
Fixed by 340ed03aba535923bf276da0403587954691f170
